### DISCUSSION TITLE
Inconsistent results

### DISCUSSION THREAD

**Discussion by nuScott** [OpenStitching/stitching#261] [Q&A] [answered]
I'm attempting to stitch video together using footage shot from two cameras pointing at a football field, one to the left and one to the right. 
Following a previous discussion I've adjusted the **stitcher.py** file to use the same **camera** and **seam mask** for each set of frames.
My issue however is the results I'm getting are inconsistent and I have no clue why. I think it may have something to do with the subset as I've been able to isolate the bad results using **cv.detail.matchesGraphAsString()** where I can see the right frame before the left frame when it should be the other way around.

I've attached results and code. Please help hah

![pano1](https://github.com/user-attachments/assets/1bfc4bb9-9177-48db-af91-a2dd50dbb9a0)
![pano4](https://github.com/user-attachments/assets/accbf129-9b0c-493a-934c-cf9a91b90eaf)
[Stitching_tutorial.docx](https://github.com/user-attachments/files/18217410/Stitching_tutorial.docx)
[stitcher.docx](https://github.com/user-attachments/files/18217413/stitcher.docx)

**Comment by lukasalexanderweber** [comment]
Won't open code in a docx file

**Reply by nuScott** [reply]
#https://github.com/OpenStitching/stitching_tutorial/blob/master/docs/Stitching%20Tutorial.md

from matplotlib import pyplot as plt
from pathlib import Path
from stitching import Stitcher
from stitching.images import Images
import cv2 as cv
import os

# displaying resulting images within the notebook
def plot_image(img, figsize_in_inches=(5,5)):
    fig, ax = plt.subplots(figsize=figsize_in_inches)
    ax.imshow(cv.cvtColor(img, cv.COLOR_BGR2RGB))
    plt.show()
    
def plot_images(imgs, figsize_in_inches=(5,5)):
    fig, axs = plt.subplots(1, len(imgs), figsize=figsize_in_inches)
    for col, img in enumerate(imgs):
        axs[col].imshow(cv.cvtColor(img, cv.COLOR_BGR2RGB))
    plt.show()

#load the correct img paths
def get_image_paths(img_fol):
    return [str(path.relative_to('.')) for path in Path(img_fol).iterdir()if path.suffix == ".png"]

#####################################################################
#Importing images to be stitched and creating various sizes of them #
#####################################################################

count = 1

path = f"./test frames/frame1"
imgs = get_image_paths(path)
images = Images.of(imgs)
low_imgs = list(images.resize(Images.Resolution.LOW))
medium_imgs = list(images.resize(Images.Resolution.MEDIUM))
final_imgs = list(images.resize(Images.Resolution.FINAL))

stitcher = Stitcher(detector="orb", confidence_threshold=.5, blend_strength=20)
stitcher.stitch(final_imgs)
subset = (cv.detail.matchesGraphAsString(imgs, stitcher.matches, .5))
bad_results = []

while count <= 10:
    #print(f"frame {count}")

    path = f"./test frames/frame{count}"
    imgs = get_image_paths(path)
    images = Images.of(imgs)

    #low_imgs = list(images.resize(Images.Resolution.LOW))
    #medium_imgs = list(images.resize(Images.Resolution.MEDIUM))
    final_imgs = list(images.resize(Images.Resolution.FINAL))
    
    panorama = stitcher.stitch(final_imgs)
    c_subset = cv.detail.matchesGraphAsString(imgs, stitcher.matches, .5)
    #print(c_subset)
    
    if c_subset.startswith("""graph matches_graph{\n"rframe"""):
        bad_results.append(count)
        print(f"Pano {count} Bad result!!")
    else:
        print(f"Pano {count} Good result!")
    
    cv.imwrite(os.path.join(path, f"pano{count}.png"), panorama) 

    #plot_image(panorama, (10,10))
    count = count + 1

print(bad_results)

**Reply by nuScott** [reply]
def stitch(self, images, feature_masks=[]):
        self.images = Images.of(
            images, self.medium_megapix, self.low_megapix, self.final_megapix
        )


        if not self.cameras_registered:
            imgs = self.resize_medium_resolution()
            features = self.find_features(imgs, feature_masks)
            self.features = features
            matches = self.match_features(features)
            self.matches = matches
            imgs, features, matches = self.subset(imgs, features, matches)
            cameras = self.estimate_camera_parameters(features, matches)
            cameras = self.refine_camera_parameters(features, matches, cameras)
            cameras = self.perform_wave_correction(cameras)
            self.estimate_scale(cameras)
            self.cameras = cameras
            self.cameras_registered = True

        imgs = self.resize_low_resolution()
        imgs, masks, corners, sizes = self.warp_low_resolution(imgs, self.cameras)
        self.prepare_cropper(imgs, masks, corners, sizes)
        imgs, masks, corners, sizes = self.crop_low_resolution(
            imgs, masks, corners, sizes
        )
        self.estimate_exposure_errors(corners, imgs, masks)

        if not self.seam_masks_registered:
            self.seam_masks = self.find_seam_masks(imgs, corners, masks)
            self.seam_masks_registered = True
            
        imgs = self.resize_final_resolution()
        imgs, masks, corners, sizes = self.warp_final_resolution(imgs, self.cameras)
        imgs, masks, corners, sizes = self.crop_final_resolution(
            imgs, masks, corners, sizes
        )
        
        self.set_masks(masks)
        imgs = self.compensate_exposure_errors(corners, imgs)
        self.seam_maskss = self.resize_seam_masks(self.seam_masks)

        self.initialize_composition(corners, sizes)
        self.blend_images(imgs, self.seam_maskss, corners)
        return self.create_final_panorama()

**Reply by lukasalexanderweber** [reply]
In the stitching tutorial, each stitch generates its own cameras (you use the default stitcher). You have to use the VideoStitcher for fixed cameras.

**Reply by nuScott** [reply]
I adjusted the code from the stitching tutorial so that only one camera is used for all stitches. I used the code you suggested in discussion #148 and it worked. 

The issue I'm having is that the results aren't consistent. Some are good and some are bad..

**Reply by nuScott** [reply]
VideoStitcher?

**Reply by lukasalexanderweber** [reply]
Where do you adjust the code? You import Stitcher and use it, I can't see where you adjust it

**Reply by nuScott** [reply]
in the stitcher.py file

    if not self.cameras_registered:
        imgs = self.resize_medium_resolution()
        features = self.find_features(imgs, feature_masks)
        self.features = features
        matches = self.match_features(features)
        self.matches = matches
        imgs, features, matches = self.subset(imgs, features, matches)
        cameras = self.estimate_camera_parameters(features, matches)
        cameras = self.refine_camera_parameters(features, matches, cameras)
        cameras = self.perform_wave_correction(cameras)
        self.estimate_scale(cameras)
        self.cameras = cameras
        self.cameras_registered = True

**Reply by nuScott** [reply]
Once a camera is created the first time the stitch method is called, it's always used

**Reply by lukasalexanderweber** [reply]
But how should the tutorial be aware of your changes in stitcher.py? Do you build an install it again with pip?

**Reply by nuScott** [reply]
I just click **save** on the stitcher.py file after making changes. The changes are then reflected on the Stitching_tutorial.py file

**Reply by nuScott** [reply]
I'm using vscode btw

**Reply by lukasalexanderweber** [reply]
Can you add a print("this is from my Code") to see if its reflected?

**Reply by nuScott** [reply]
I added it just before the return statement

<img width="641" alt="Screenshot 2024-12-21 at 10 30 29" src="https://github.com/user-attachments/assets/a7b8a95f-95a9-4b87-9e15-cde60264ce02" />
<img width="799" alt="Screenshot 2024-12-21 at 10 32 17" src="https://github.com/user-attachments/assets/0731a179-6eca-4025-8fa6-c91dbc7abfef" />

**Reply by lukasalexanderweber** [reply]
Can you add another print within the registration? We should only see it once.

**Reply by nuScott** [reply]
<img width="731" alt="Screenshot 2024-12-21 at 11 06 56" src="https://github.com/user-attachments/assets/650d1fc7-910b-44db-a7f9-96623cfd3d4b" />
<img width="652" alt="Screenshot 2024-12-21 at 11 06 21" src="https://github.com/user-attachments/assets/7d1c22cc-789f-4aed-82d5-420b7ddd87e0" />

**Reply by lukasalexanderweber** [reply]
Btw are you sure that you always stitch([leftFrame, rightFrame])? The order is very important. Maybe you need to adjust get_image_paths to ensure this

**Reply by lukasalexanderweber** [reply]
print(imgs) for every frame pair

**Reply by nuScott** [reply, marked as answer]
Looks like the order is changing within the while loop. Why would it do that? Maybe get_image_paths method is the issue?

<img width="624" alt="Screenshot 2024-12-21 at 11 18 16" src="https://github.com/user-attachments/assets/973d96c3-fad2-4976-8e19-85682d0e0362" />
<img width="539" alt="Screenshot 2024-12-21 at 11 17 42" src="https://github.com/user-attachments/assets/35319230-200f-4538-8f53-92aabd693e1a" />

**Comment by nuScott** [comment]
I managed to get the results I desired thanks to you, cheers pal

**Reply by lukasalexanderweber** [reply]
Great, good luck with your project!
